File: src/main/java/com/baulsupp/oksocial/commands/SimpleCommand.java
package com.baulsupp.oksocial.commands;

import com.baulsupp.oksocial.output.util.UsageException;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import okhttp3.OkHttpClient;
import okhttp3.Request;

public class SimpleCommand implements ShellCommand {
  private String name;
  private String prefix;
  private String authName;

  public SimpleCommand(String name, String prefix, String authName) {
    this.name = name;
    this.prefix = prefix;
    this.authName = authName;
  }

  @Override public String name() {
    return name;
  }

  public String mapUrl(String url) {
    return prefix + url;
  }

  @Override public List<Request> buildRequests(OkHttpClient clientBuilder,
      Request.Builder requestBuilder, List<String> urls) {
    try {
      return urls.stream().map(u -> requestBuilder.url(mapUrl(u)).build()).collect(
          Collectors.toList());
    } catch (IllegalArgumentException iae) {
      throw new UsageException(iae.getMessage());
    }
  }

  @Override public Optional<String> authenticator() {
    return Optional.ofNullable(authName);
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/commands/SimpleCommand.kt
package com.baulsupp.oksocial.commands

import com.baulsupp.oksocial.kotlin.request
import com.baulsupp.oksocial.output.UsageException
import okhttp3.OkHttpClient
import okhttp3.Request

open class SimpleCommand(val name: String, private val prefix: String, val authenticator: String?) : ShellCommand {

  override fun name(): String {
    return name
  }

  override fun buildRequests(client: OkHttpClient,
                             arguments: List<String>): List<Request> {
    return try {
      arguments.map { u -> request(prefix + u) }
    } catch (iae: IllegalArgumentException) {
      throw UsageException(iae.message.orEmpty())
    }
  }
}
